int32 i
int32 k
int32 j
int32 index
float32 val

